### PR TITLE
Support more improvement in optional fields

### LINKED ISSUE #36021: More improvements on working with optional fields
Points: 3
Spec Issue: https://github.com/ballerina-platform/ballerina-spec/issues/1079

### PR DESCRIPTION
**Pull request by KavinduZoysa**
## Purpose

Fixes #36021

### CODE REVIEW

**Comment by KavinduZoysa** [comment]
Few tests are failing due to https://github.com/ballerina-platform/ballerina-lang/issues/37709. I will fix this

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/statements/vardeclr/ModuleRecordVariableTest.java [hunk 1]
```diff
@@ -84,13 +85,10 @@ public void testModuleLevelRecordVarDeclNegetive() {
...
-                "invalid field binding pattern; can only bind required fields", 75, 28);
+                "incompatible types: expected 'record {| int id; int age?; anydata...; |}', " +
+                        "found 'record {| int id; int age?; anydata...; |}?'", 91, 31);
         validateError(compileResultNegetive, index++,
-                "invalid field binding pattern; can only bind required fields", 77, 39);
-        validateError(compileResultNegetive, index++,
-                "invalid field binding pattern; can only bind required fields", 91, 22);
-        validateError(compileResultNegetive, index++,
-                "invalid field binding pattern; can only bind required fields", 91, 43);
+                "incompatible types: expected 'record {| int b?; anydata...; |}[1]', found 'record {| int b?; anydata...; |}[1]?'", 101, 12);
```
Comment (gimantha) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/statements/vardeclr/ModuleRecordVariableTest.java:91]:
```suggestion
                "incompatible types: expected 'record {| int b?; anydata...; |}[1]', " + 
                "found 'record {| int b?; anydata...; |}[1]?'", 101, 12);
```

Comment (KavinduZoysa) [reply tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/statements/vardeclr/ModuleRecordVariableTest.java]: Changed

File: tests/jballerina-unit-test/src/test/resources/test-src/statements/vardeclr/module_record_var_decl.bal [hunk 2]
```diff
@@ -181,6 +181,30 @@ type Employee record {
...
+int? x2 = ();
+int? y2 = 5;
+Topt topt3 = {x: x2, y: y2};
+
+function testOptionalFieldAssignment() {
+    assertEquality(2, x);
+    assertEquality(4, y);
+    assertEquality(2, x1);
+    assertTrue(y1 is ());
+    assertTrue(topt3.x is ());
```
Comment (gimantha) [inline tests/jballerina-unit-test/src/test/resources/test-src/statements/vardeclr/module_record_var_decl.bal:204]: Should we not check if the field is not there at all?

File: tests/jballerina-unit-test/src/test/resources/test-src/statements/variabledef/record-variable-definition-stmt.bal [hunk 3]
```diff
@@ -468,10 +468,145 @@ function testRestFieldResolving() {
...
+}
+
+function testRecordDefinitionWithOptionalFields7() {
+    int x = 2;
+    int? y = ();
+    Topt1 topt = {x, y};
+    assertEquality(topt.x, 2);
+    assertTrue(topt.y is ());
+}
+
```
Comment (gimantha) [inline tests/jballerina-unit-test/src/test/resources/test-src/statements/variabledef/record-variable-definition-stmt.bal:601]: Shall we add some tests to check whether the query cases now works same for optional fields and nillable types as mentioned in https://github.com/ballerina-platform/ballerina-spec/issues/1079?

Comment (KavinduZoysa) [reply tests/jballerina-unit-test/src/test/resources/test-src/statements/variabledef/record-variable-definition-stmt.bal:601]: Added

**Comment by MaryamZi** [comment]
Shouldn't the behaviour be similar with member-access-lvexpr? May have to check in the spec issue.

```ballerina
type Topt record {
    int x?;
    int y?;
};

public function main() {
    Topt t = {};
    t["y"] = (); // an error atm: incompatible types: expected 'int', found '()'
}
```

**Comment by KavinduZoysa** [comment]
> Shouldn't the behaviour be similar with member-access-lvexpr? May have to check in the spec issue.
> 
> This results in an error
> 
> ```ballerina
> type Topt record {
>     int x?;
>     int y?;
> };
> 
> public function main() {
>     Topt t = {};
>     t["y"] = (); // an error atm: incompatible types: expected 'int', found '()'
> }
> ```
> 
> But this removes the field.
> 
> ```ballerina
> import ballerina/io;
> 
> type Topt record {
>     int x?;
>     int y?;
> };
> 
> public function main() {
>     Topt t = {y: 1};
>     io:println(t);
>     string s = "y"; // {"y":1}
>     t[s] = ();
>     io:println(t); // {}
> }
> ```

Fix this and added tests

**Comment by KavinduZoysa** [comment]
Since this approach is not correct, created a new PR https://github.com/ballerina-platform/ballerina-lang/pull/37883, hence closing this

**Pull request closed without merging**